<a href="https://colab.research.google.com/github/Aissa-C/Anal-tica/blob/main/Naive_Bayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Modelo de PreAprobación - Naive Bayes**
0. Se procede con la carga de las librerías de trabajo

In [ ]:
import numpy as np    #Funciones matemáticas
import pandas as pd   #Comunicación con Excel
from sklearn.naive_bayes import GaussianNB   #Función Gaussiana
from sklearn.metrics import confusion_matrix   #Métrica de medida

1. Se cargan los datos

In [ ]:
nxl= '/content/0. SolicitantesCrédito(USD)_NaiveBayes(L).xlsx'
XDB= pd.read_excel(nxl,sheet_name=1)
XDB.dropna()  #Eliminar datos faltantes, celdas vacías

#Seleccionamos las variables de interés
XDB=XDB[['Edad', 'Ingresos', 'Egresos', 'Cuota (COP)', 'PreApr']]

#Datos de entrada
XD= np.array(XDB[['Edad', 'Ingresos', 'Egresos', 'Cuota (COP)']])
#Datos de salida - Valor deseado - qué quiero que aprenda el modelo
yd= np.array(XDB[['PreApr']])


2. Implementamos Modelo

In [ ]:
mnb= GaussianNB()  #Llamamos el modelo
mnb.fit(XD,yd)
ydp= mnb.predict(XD)  #Muestra los datos de salida - pero que él obtuvo

#Dígame las medias de las variables para cada categoría - PreApr/PreNeg
u= mnb.theta_   #Se estiman las medias
sigma= np.sqrt(mnb.var_)   #Se estiman las varianzas

#Las variaciones de las variables
LI= u-sigma
LS= u+sigma

print("La media de las variables por categoría es:")
print(u)

print("El límite inferior de la variación por categoría es:")
print(LI)

print("El límite superior de la variación por categoría es:")
print(LS)


#Métrica de Evaluación -  cómo se comportó el modelo identificando la PreApr/PreNeg
cm= confusion_matrix(yd,ydp)

La media de las variables por categoría es:
[[  40.98411626  431.42680644  401.54995644   16.88529983]
 [  41.10752688 1029.98947251  900.19312144   43.3269839 ]]
El límite inferior de la variación por categoría es:
[[ 28.77108358 105.62889147  58.77709557 -12.88412769]
 [ 28.85646196 557.55604444 374.75974307 -21.65436704]]
El límite superior de la variación por categoría es:
[[  53.19714893  757.22472141  744.32281731   46.65472735]
 [  53.35859181 1502.42290058 1425.62649981  108.30833484]]


/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


3. Se procede con la evaluación del modelo

In [ ]:
VN= cm[0,0]  #Verdaderos Negativos
FP= cm[0,1]  #Falsos Positivos
FN= cm[1,0]  #Falsos Negativos
VP= cm[1,1]  #Verdaderos positivos

Exactitud= (VP+VN)/len(XD)
print("La exactitud es:", Exactitud)  #Comportamiento General

TE= (FP+FN)/len(XD)
print("La tasa de error es:", TE)   #Porcentaje de equivocaciones

Sensibilidad= VP/(VP+FP)
print("La sensibilidad es:", Sensibilidad)  #Cómo se comportó frente a los positivos

Especificidad= VN/(VN+FN)
print("La especificidad es:", Especificidad)  #Cómo se comportó frente a los negativos

Precision= VP/(VP+FN)
print("La precisión es:", Precision)   #Cómo le pegó a los positivos

PredNeg= VN/(VN+FP)
print("La probabilidad de negativo es:", PredNeg)   #Cómo le pegó a los negativos

La exactitud es: 0.7281752824375214
La tasa de error es: 0.2718247175624786
La sensibilidad es: 0.8102539530426449
La especificidad es: 0.6825565912117177
La precisión es: 0.5865417967395075
La probabilidad de negativo es: 0.8661710037174721


4. Qué pasa con una sola persona

In [ ]:
Xun= np.array([[38, 1356, 1686, 18]])   #Esta es la persona del Excel
ydun= mnb.predict(Xun)
print(ydun)

[1]


Análisis de Resultados:
De los resultados arrojados por el modelo se puede observar que los PreApr en general presentan edades (), ingresos (), egresos(), y cuotas() mayores a los PreNeg. Con respecto a la métrica se puede observar que el modelo funcionó mejor clasificando los positivos, como lo dice el indicador de sensibilidad el cual alcanzó el 81% de certezaz. De manera genreal, se puede observar que el modelo logró una exactitud que se ubicó por encima del 70% en promedio, y una tasa error cercana al 30% en promedio. De manera general, esta exactitud lo convierte en un buen modelo. Con respecto al solicitante de crédito tal y tal, este pertenece a la categoría de aprobado.